File: strategy-engine/src/main/java/com/quantstream/strategy/strategies/StochasticStrategy.java
package com.quantstream.strategy.strategies;

import com.quantstream.strategy.framework.TradingStrategy;
import com.quantstream.strategy.model.Signal;
import com.quantstream.strategy.utils.IndicatorUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Stochastic Oscillator Strategy
 *
 * Type: Momentum / Oscillator
 *
 * Components:
 * - %K: Fast stochastic (current momentum)
 * - %D: Slow stochastic (3-period SMA of %K)
 *
 * Logic:
 * - %K crosses above %D in oversold zone (<20) → BUY
 * - %K crosses below %D in overbought zone (>80) → SELL
 *
 * Performance:
 * - Best in ranging markets with clear support/resistance
 * - Poor in strong trends (stays in extreme zones)
 */
@Component
public class StochasticStrategy implements TradingStrategy {

    private static final Logger log = LoggerFactory.getLogger(StochasticStrategy.class);

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Autowired
    private IndicatorUtils indicators;

    private static final int STOCH_PERIOD = 14;
    private static final int D_PERIOD = 3;
    private static final double OVERSOLD = 20.0;
    private static final double OVERBOUGHT = 80.0;

    // Track %K history to calculate %D (SMA of %K)
    private final Map<String, List<Double>> percentKHistory = new HashMap<>();

    @Override
    public String getName() {
        return "STOCHASTIC";
    }

    @Override
    public int getRequiredHistoryDays() {
        return STOCH_PERIOD;
    }

    @Override
    public Signal analyze(String symbol) {
        try {
            // Step 1: Query historical prices
            List<Double> prices = queryPrices(symbol, STOCH_PERIOD);

            // Step 2: Validate data
            if (prices.size() < STOCH_PERIOD) {
                log.debug("Not enough data for {}: {} ticks (need {})",
                    symbol, prices.size(), STOCH_PERIOD);
                return null;
            }

            // Step 3: Calculate %K
            IndicatorUtils.Stochastic stoch = indicators.calculateStochastic(prices, STOCH_PERIOD);
            double percentK = stoch.getPercentK();

            // Step 4: Update %K history
            List<Double> kHistory = percentKHistory.computeIfAbsent(symbol, k -> new ArrayList<>());
            kHistory.add(0, percentK);  // Add to front

            // Keep only what we need for %D
            if (kHistory.size() > D_PERIOD + 5) {
                kHistory.remove(kHistory.size() - 1);
            }

            // Step 5: Calculate %D (need at least 3 %K values)
            if (kHistory.size() < D_PERIOD) {
                log.debug("Building %K history for {}: {}/{}",
                    symbol, kHistory.size(), D_PERIOD);
                return null;
            }

            double percentD = kHistory.stream().limit(D_PERIOD)
                .mapToDouble(Double::doubleValue)
                .average()
                .orElse(0.0);

            // Step 6: Check for previous values to detect crossover
            if (kHistory.size() < D_PERIOD + 1) {
                log.debug("First complete run for {}, initializing state", symbol);
                return null;
            }

            // Get previous %K and %D
            double prevPercentK = kHistory.get(1);
            double prevPercentD = kHistory.subList(1, D_PERIOD + 1).stream()
                .mapToDouble(Double::doubleValue)
                .average()
                .orElse(0.0);

            // Step 7: Detect crossovers in extreme zones
            Signal signal = null;

            // Bullish: %K crosses above %D in oversold zone
            if (percentK > percentD && prevPercentK <= prevPercentD && percentK < OVERSOLD + 10) {
                signal = new Signal(
                    symbol,
                    "BUY",
                    getName(),
                    calculateConfidence(percentK, true),
                    Instant.now()
                );
                log.info("Stochastic bullish crossover: {} (%K={:.2f}, %D={:.2f})",
                    symbol, percentK, percentD);
            }

            // Bearish: %K crosses below %D in overbought zone
            else if (percentK < percentD && prevPercentK >= prevPercentD && percentK > OVERBOUGHT - 10) {
                signal = new Signal(
                    symbol,
                    "SELL",
                    getName(),
                    calculateConfidence(percentK, false),
                    Instant.now()
                );
                log.info("Stochastic bearish crossover: {} (%K={:.2f}, %D={:.2f})",
                    symbol, percentK, percentD);
            }

            return signal;

        } catch (Exception e) {
            log.error("Stochastic strategy failed for {}: {}", symbol, e.getMessage());
            return null;
        }
    }

    /**
     * Query recent prices from QuestDB.
     */
    private List<Double> queryPrices(String symbol, int limit) {
        String sql = "SELECT price FROM ticks WHERE symbol = ? ORDER BY timestamp DESC LIMIT ?";

        return jdbcTemplate.query(
            sql,
            (rs, rowNum) -> rs.getDouble("price"),
            symbol,
            limit
        );
    }

    /**
     * Calculate confidence based on position in extreme zone.
     *
     * More extreme = higher confidence.
     */
    private double calculateConfidence(double percentK, boolean isBuy) {
        double baseConfidence = 0.70;

        if (isBuy) {
            // Lower %K in oversold zone = higher confidence
            double extremity = (OVERSOLD - percentK) / OVERSOLD;
            return Math.min(baseConfidence + (Math.max(0, extremity) * 0.20), 0.90);
        } else {
            // Higher %K in overbought zone = higher confidence
            double extremity = (percentK - OVERBOUGHT) / (100 - OVERBOUGHT);
            return Math.min(baseConfidence + (Math.max(0, extremity) * 0.20), 0.90);
        }
    }
}

File: backtester/app/strategies/stochastic_strategy.py
"""
Stochastic Oscillator Strategy

Buys when %K crosses above %D in oversold zone.
Sells when %K crosses below %D in overbought zone.

Matches Phase 2: StochasticStrategy.java
"""

from typing import Optional
import pandas as pd
from app.strategies.base_strategy import BaseStrategy
from app.models.signal import Signal
from app.core.indicators import calculate_stochastic


class StochasticStrategy(BaseStrategy):
    """
    Stochastic Oscillator Strategy.

    Parameters:
        - %K Period: 14
        - %D Period: 3 (SMA of %K)
        - Oversold: 20
        - Overbought: 80

    Logic:
        - BUY: %K crosses above %D AND %K < 30 (oversold zone)
        - SELL: %K crosses below %D AND %K > 70 (overbought zone)

    Confidence:
        - Based on extremity (how oversold/overbought)
        - Range: 0.75 to 0.90
    """

    STOCH_PERIOD = 14
    D_PERIOD = 3
    OVERSOLD = 20.0
    OVERBOUGHT = 80.0

    def __init__(self):
        super().__init__(name="Stochastic")

    def reset_state(self):
        """Reset previous %K and %D values."""
        self.prev_k = None
        self.prev_d = None

    def get_required_candles(self) -> int:
        """Need 14 candles for stochastic calculation."""
        return self.STOCH_PERIOD

    def generate_signal(self, df: pd.DataFrame) -> Optional[Signal]:
        """
        Generate signal based on Stochastic crossover in extreme zones.

        Args:
            df: DataFrame with 'high', 'low', 'close' prices

        Returns:
            Signal (BUY/SELL) or None
        """
        # Calculate Stochastic
        df_with_stoch = calculate_stochastic(
            df.copy(),
            k_period=self.STOCH_PERIOD,
            d_period=self.D_PERIOD
        )

        current_k = df_with_stoch['stoch_k'].iloc[-1]
        current_d = df_with_stoch['stoch_d'].iloc[-1]

        # First run: initialize state
        if self.prev_k is None:
            self.prev_k = current_k
            self.prev_d = current_d
            return None

        signal = None

        # BUY: %K crosses above %D in oversold zone
        # Use 30 threshold (OVERSOLD + 10 buffer) as in Phase 2
        if (current_k > current_d and self.prev_k <= self.prev_d and
            current_k < 30):
            confidence = self._calculate_confidence(current_k, is_buy=True)
            signal = Signal(action="BUY", confidence=confidence)

        # SELL: %K crosses below %D in overbought zone
        # Use 70 threshold (OVERBOUGHT - 10 buffer) as in Phase 2
        elif (current_k < current_d and self.prev_k >= self.prev_d and
              current_k > 70):
            confidence = self._calculate_confidence(current_k, is_buy=False)
            signal = Signal(action="SELL", confidence=confidence)

        # Update state for next call
        self.prev_k = current_k
        self.prev_d = current_d
        return signal

    def _calculate_confidence(self, k: float, is_buy: bool) -> float:
        """
        Calculate confidence based on extremity.

        More extreme %K = higher confidence.

        Args:
            k: Current %K value
            is_buy: True for BUY signal, False for SELL

        Returns:
            Confidence between 0.75 and 0.90
        """
        if is_buy:
            # More oversold (lower %K) = higher confidence
            bonus = (30 - k) / 30 * 0.15
        else:
            # More overbought (higher %K) = higher confidence
            bonus = (k - 70) / 30 * 0.15

        return min(0.90, 0.75 + bonus)
